### DISCUSSION TITLE
Ciphertext Block Size

### DISCUSSION THREAD

**Discussion by overheadhunter** [encryption-alliance/unified-vault-format#13] [Ideas]
As [already mentioned in a different discussion](https://github.com/cryptomator/new-encryption-format/discussions/5#discussioncomment-5023444), the block size (i.e. how many bytes per cipher-reinitialization) may be used to tune performance for different use cases.

Files that are mostly read/written sequentially (like video files) benefit from not needing to re-init the cipher with a new nonce every few kB. On the other hand, during read we must keep the cleartext in memory until we have verified the tag. While 4kiB seems to be a de-facto standard on block devices, in Cryptomator we found 32kiB blocks to be a reasonable compromise between memory consumption and cipher throughput, which works for most file types.

However, I don't have any data to back this claim. So we should do some testing before agreeing on one or several block sizes.

As sufficiently justified in #11, such a configuration, which affects the ciphertext file size, must be made _per vault_ (not per dir or per file).

**Comment by ncw** [comment]
I used 64k in rclone crypt. I chose that after a few hours of trying different values trying to come up with a value that gave max throughput for minimum memory use.

I made crypt 7 years ago nearly and I'd probably choose a larger value today.

Though choosing a larger value means more work when randomly seeking through a file (eg reading the metadata from an audio/video file) so there are lots of tradeoffs.

**Reply by rfjakob** [reply]
Hmm, how are block size and memory usage related? The requests you get via FUSE are usually larger ( 128kiB to up to 1 MiB on latest kernels), so this should determine the memory usage.

**Reply by ncw** [reply]
You need to keep a block or two in memory one to fill up with encrypted data and another to decrypt into. This needs to be kept around until it is emptied if the requests don't like up with the block size (quite likely for non-FUSE applications).

**Reply by overheadhunter** [reply]
Btw the iOS file provider extensions have very strict memory limits. Sub-MiB blocks _should_ be fine, but I am not an expert in this field. @tobihagemann I believe we had issues with a memory-hard KDF but do we know the exact limit when the OS is going kill the process?

**Reply by tobihagemann** [reply]
In the past, the memory limit for FileProvider extensions was at 15 MB but it seems to have been increased to 20 MB about last year.

We were never able to determine how much MB we _really_ have since attaching a debugger causes a memory increase, which would increase the probability of memory limit-related crashes. We just know that our current configurations and optimizations work. 😄

Edit: Memory-hard KDF ist not an issue because the FileProviderUI, which performs the unlock process, has a memory limit of 200 MB, I think.

**Comment by rfjakob** [comment]
I don't know what kind of throughput you would like to target, but gocryptfs pushes close to 1 GB/s with 4kiB blocks in AES-GCM-256:

```
$ ./gocryptfs -speed
gocryptfs v2.3.1; go-fuse v2.1.1-0.20221117175120-915cf5413cde; 2023-03-07 go1.19.6 linux/amd64
cpu: Intel(R) Core(TM) i5-3470 CPU @ 3.20GHz; with AES acceleration
AES-GCM-256-OpenSSL       	 700.14 MB/s
AES-GCM-256-Go            	 988.40 MB/s	(selected in auto mode)
AES-SIV-512-Go            	 158.32 MB/s
XChaCha20-Poly1305-OpenSSL	 570.28 MB/s
XChaCha20-Poly1305-Go     	 835.31 MB/s	(selected in auto mode)
```

Same benchmark with 64 kiB blocks:

```
$ ./gocryptfs -speed
gocryptfs v2.3.1-dirty; go-fuse v2.1.1-0.20221117175120-915cf5413cde; 2023-03-07 go1.19.6 linux/amd64
cpu: Intel(R) Core(TM) i5-3470 CPU @ 3.20GHz; with AES acceleration
AES-GCM-256-OpenSSL       	1035.58 MB/s
AES-GCM-256-Go            	1029.69 MB/s	(selected in auto mode)
AES-SIV-512-Go            	 176.55 MB/s
XChaCha20-Poly1305-OpenSSL	 870.14 MB/s
XChaCha20-Poly1305-Go     	 897.43 MB/s	(selected in auto mode)
```

Edit: Whole block size range now via https://github.com/rfjakob/gocryptfs/commit/d74cf7c72337e7b6d4c57d357e855b19fe1f5c9a  & https://github.com/rfjakob/gocryptfs/commit/b370325ccf8a15d3b58418d85fd22e32e2aeb2fc:

```
BenchmarkGoGCMBlockSize/16-4      	 5499200	       219.7 ns/op	  72.83 MB/s
BenchmarkGoGCMBlockSize/32-4      	 4497284	       266.2 ns/op	 120.22 MB/s
BenchmarkGoGCMBlockSize/64-4      	 3296336	       363.4 ns/op	 176.10 MB/s
BenchmarkGoGCMBlockSize/128-4     	 4204794	       285.5 ns/op	 448.36 MB/s
BenchmarkGoGCMBlockSize/256-4     	 2928472	       409.7 ns/op	 624.83 MB/s
BenchmarkGoGCMBlockSize/512-4     	 1825164	       658.0 ns/op	 778.09 MB/s
BenchmarkGoGCMBlockSize/1024-4    	 1000000	      1151 ns/op	 889.98 MB/s
BenchmarkGoGCMBlockSize/2048-4    	  560275	      2135 ns/op	 959.47 MB/s
BenchmarkGoGCMBlockSize/4096-4    	  291906	      4099 ns/op	 999.28 MB/s
BenchmarkGoGCMBlockSize/8192-4    	  148916	      8033 ns/op	1019.83 MB/s
BenchmarkGoGCMBlockSize/16384-4   	   75337	     15911 ns/op	1029.75 MB/s
BenchmarkGoGCMBlockSize/32768-4   	   37912	     31651 ns/op	1035.30 MB/s
BenchmarkGoGCMBlockSize/65536-4   	   19000	     64287 ns/op	1019.43 MB/s
BenchmarkGoGCMBlockSize/131072-4  	    9225	    127636 ns/op	1026.92 MB/s
BenchmarkGoGCMBlockSize/262144-4  	    4752	    252300 ns/op	1039.02 MB/s
BenchmarkGoGCMBlockSize/524288-4  	    2377	    504612 ns/op	1038.99 MB/s
BenchmarkGoGCMBlockSize/1048576-4 	    1183	   1011637 ns/op	1036.51 MB/s
```

**Reply by overheadhunter** [reply]
The numbers suggest that while OpenSSL experiences a speedup, the Go variant hits some other throughput limiting bottleneck. At least in this specific test on this machine.
I would therefore not conclude that larger block size doesn't impact Go's performance as well.

**Reply by rfjakob** [reply]
Calling openssl from Go is expensive, which is why there's some speedup. But let's take Go out of the equation and see what straight openssl says:

```
$ openssl speed -evp aes-256-gcm
Doing AES-256-GCM for 3s on 16 size blocks: 40668727 AES-256-GCM's in 2.99s
Doing AES-256-GCM for 3s on 64 size blocks: 29531671 AES-256-GCM's in 2.99s
Doing AES-256-GCM for 3s on 256 size blocks: 11154802 AES-256-GCM's in 2.99s
Doing AES-256-GCM for 3s on 1024 size blocks: 3132515 AES-256-GCM's in 2.99s
Doing AES-256-GCM for 3s on 8192 size blocks: 398071 AES-256-GCM's in 2.97s
Doing AES-256-GCM for 3s on 16384 size blocks: 201568 AES-256-GCM's in 2.99s
[...]
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-256-GCM     217625.29k   632116.03k   955059.97k  1072807.81k  1097979.00k  1104511.74k
```

```
$ openssl speed -evp aes-256-gcm -bytes 65536
[...]
AES-256-GCM    1111942.08k
```

```
$ openssl speed -evp aes-256-gcm -bytes 1048576
[...]
type        1048576 bytes
AES-256-GCM    1112402.37k
```

Looking at the numbers, we should use at least 1k-sized blocks for best performance.

However, openssl is the best-optimzed library out there, and things may look different in Java land.

**Reply by ncw** [reply]
I wrote a little benchmark for nacl secretbox as implemented by  golang.org/x/crypto/nacl/secretbox (which is what rclone crypt uses)

This is encrypting a block of the given size.

64k block sizes seem about right for maximising throughput.

```
BenchmarkSecretBox/Block1024-16         	  385686	      3062 ns/op	 334.47 MB/s
BenchmarkSecretBox/Block2048-16         	  229050	      5278 ns/op	 388.04 MB/s
BenchmarkSecretBox/Block4096-16         	  116938	      9723 ns/op	 421.29 MB/s
BenchmarkSecretBox/Block8192-16         	   64960	     18403 ns/op	 445.15 MB/s
BenchmarkSecretBox/Block16384-16        	   32606	     36090 ns/op	 453.98 MB/s
BenchmarkSecretBox/Block32768-16        	   16888	     72415 ns/op	 452.50 MB/s
BenchmarkSecretBox/Block65536-16        	    8396	    141821 ns/op	 462.10 MB/s
BenchmarkSecretBox/Block131072-16       	    3550	    285877 ns/op	 458.49 MB/s
BenchmarkSecretBox/Block262144-16       	    1918	    568580 ns/op	 461.05 MB/s
BenchmarkSecretBox/Block524288-16       	     968	   1135874 ns/op	 461.57 MB/s
BenchmarkSecretBox/Block1048576-16      	     525	   2255741 ns/op	 464.85 MB/s
BenchmarkSecretBox/Block2097152-16      	     258	   4549934 ns/op	 460.92 MB/s
BenchmarkSecretBox/Block4194304-16      	     130	   9122118 ns/op	 459.79 MB/s
BenchmarkSecretBox/Block8388608-16      	      66	  18207232 ns/op	 460.73 MB/s
BenchmarkSecretBox/Block16777216-16     	      32	  36559088 ns/op	 458.91 MB/s
BenchmarkSecretBox/Block33554432-16     	      15	  72426491 ns/op	 463.29 MB/s
BenchmarkSecretBox/Block67108864-16     	       7	 146642101 ns/op	 457.64 MB/s
BenchmarkSecretBox/Block134217728-16    	       4	 290694784 ns/op	 461.71 MB/s
BenchmarkSecretBox/Block268435456-16    	       2	 669972322 ns/op	 400.67 MB/s
BenchmarkSecretBox/Block536870912-16    	       1	1348009474 ns/op	 398.27 MB/s
BenchmarkSecretBox/Block1073741824-16   	       1	2663714069 ns/op	 403.10 MB/s
BenchmarkSecretBox/Block2147483648-16   	       1	5349235855 ns/op	 401.46 MB/s
```

**Reply by overheadhunter** [reply]
Honestly, the speedup between 4k and 128M looks negligible and also there isn't any clear trend (all 440 MiB/s ±5% probably just noise due to memory alignment effects). If throughput was the only factor, we should use the fastest, but as mentioned before, we need to keep in mind that larger blocks also means larger overhead for random access, negatively affecting performance.

Let's try to also get some numbers for WebCrypto API, iOS (CryptoKit/CommonCrypto) and Java (Desktop/Android) if there is a more significant speedup for certain configurations. But I guess the ideal block size doesn't need to be set in stone today, we can make a final decision right before publishing 1.0.0 of this spec.

**Reply by rfjakob** [reply]
Case for 4k: That's the page size sqlite uses per default. I imagine users may have sqlite databases in their vaults (a lot of apps use sqlite internally).

**Reply by overheadhunter** [reply]
I guess it really depends on the kind of data users are storing. In other words: Whatever we agree on will be the wrong decision for somebody. So I guess we should make it configurable during vault creation.

**Reply by rfjakob** [reply]
Making it configurable would be necessary if we cannot find a block size that works well for all use cases. I'm still optimistic that we can.

**Reply by ncw** [reply]
Is this 4k of user data in a block, making the actual blocks slightly bigger than 4k or a 4k block size making the user data within slightly smaller than 4k?

(I am assuming each block has a crypto authenticator in it.)

Either way I think that would make smaller blocks less attractive.

**Reply by overheadhunter** [reply]
> if we cannot find a block size that works well for all use cases

Well I suppose any block size that doesn't hit in-mem boundaries on restricted devices such as browsers or mobile devices would "work". But there is no one-size-fits-all solution in regards to efficiency. Take the dilemma between random access (small blocks → faster seek) and sequential access (larger blocks → fewer cipher reinit).

So "works well for all use cases" would always be a compromise. As mentioned before, we found 32k to be acceptable in Cryptomator (knowing that it doesn't offer ideal performance in all cases). And I would argue that implementors of this standard are still free to impose some default value during vault creation, as long as they correctly store it in the vault settings and support opening a vault with a different block size.

**Comment by rfjakob** [comment]
Yes the idea is 4096 bytes of plaintext. I expect 32 bytes of overhead per
block, or 0.8%.

**Reply by ncw** [reply]
That is going to make writing encrypted blocks to disk / storing them in memory less efficient than it could be due to page size effects.

Taking the overhead within the block pushes the problem back to the client.

I chose to take the overhead within the block with rclone as I decided that I liked my encrypted blocks to be exactly 64k long to make them efficient to store on disk / send over the network / etc.

In practice this probably makes very little difference I expect!

**Reply by rfjakob** [reply]
Huh, I did not think about aligning the ciphertext size, but yeah, could make sense! Adds a 2nd dimension to the question.
The file header will still cause misalignment on disk, though.
